examples/samplers/simulated_annealing_sampler.py

```diff
@@ -0,0 +1,146 @@
+"""
+Optuna example that implements a user-defined relative sampler based on Simulated Annealing
+algorithm. Please refer to https://en.wikipedia.org/wiki/Simulated_annealing for Simulated
+Annealing itself.
+
+Note that this implementation isn't intended to be used for production purposes and
+has the following limitations:
+- The sampler only supports `UniformDistribution` (i.e., `Trial.suggest_uniform` method).
+- The implementation prioritizes simplicity over optimization efficiency.
+
+You can run this example as follows:
+    $ python simulated_annealing_sampler.py
+
+"""
+
+import numpy as np
+
+import optuna
+from optuna import distributions
+from optuna.samplers import BaseSampler
+from optuna import structs
+
+if optuna.types.TYPE_CHECKING:
+    from typing import Any  # NOQA
+    from typing import Dict  # NOQA
+    from typing import Optional  # NOQA
+
+    from optuna.distributions import BaseDistribution  # NOQA
+    from optuna.structs import FrozenTrial  # NOQA
+    from optuna.study import InTrialStudy  # NOQA
+
+COOLDOWN_FACTOR = 0.9
```

Comment:
Any reason why `COOLDOWN_FACTOR` and `NEIGHBOR_RANGE_FACTOR` are constants while `temperature` is the constructor's argument? We could put all of them into one place.

Reply:
There is no strong reason.
I removed those constants and added `cooldown_factor` and `neighbor_range_factor` arguments to the constructor of `SimulatedAnnealingSampler` (https://github.com/pfnet/optuna/pull/446/commits/565cc42811d5178cc17451c845864477161f61b4).
Thank you for pointing it out.